src/lib.rs

```diff
@@ -1,49 +1,163 @@
+            .map(|v| {
+                opentelemetry_semantic_conventions::resource::SERVICE_NAME.string(v)
+            });
+        let service_version = std::env::var("OTEL_SERVICE_VERSION")
+            .or_else(|_| std::env::var("SERVICE_VERSION"))
+            .or_else(|_| std::env::var("APP_VERSION"))
+            .ok()
+            .or_else(|| Some(self.fallback_service_version.to_string()))
+            .map(|v| {
+                opentelemetry_semantic_conventions::resource::SERVICE_VERSION.string(v)
+            });
+        Resource::new(vec![service_name, service_version].into_iter().flatten())
+    }
+}
+
+pub fn build_logger_text<S>(
+    log_level: tracing::Level,
+) -> Box<dyn Layer<S> + Send + Sync + 'static>
+where
+    S: Subscriber + for<'a> LookupSpan<'a>,
+{
+    if cfg!(debug_assertions) {
+        Box::new(
+            tracing_subscriber::fmt::layer()
+                .pretty()
+                .with_line_number(true)
+                .with_thread_names(true)
+                .with_timer(tracing_subscriber::fmt::time::uptime())
+                .with_span_events(FmtSpan::NEW | FmtSpan::CLOSE)
+                .with_writer(std::io::stdout.with_max_level(log_level)),
+        )
+    } else {
+        Box::new(
+            tracing_subscriber::fmt::layer()
+                .json()
+                .with_timer(tracing_subscriber::fmt::time::uptime())
+                .with_span_events(FmtSpan::NEW | FmtSpan::CLOSE)
+                .with_writer(std::io::stdout.with_max_level(log_level)),
+        )
+    }
+}
+
 pub fn init_tracing_with_fallbacks(
     log_level: tracing::Level,
     fallback_service_name: &'static str,
     fallback_service_version: &'static str,
 ) {
-    let otel_rsrc = DetectResource::default()
-        .with_fallback_service_name(fallback_service_name)
-        .with_fallback_service_version(fallback_service_version)
-        .build();
+    let otel_rsrc =
+        DetectResource::new(fallback_service_name, fallback_service_version).build();
     let otel_tracer =
         otlp::init_tracer(otel_rsrc, otlp::identity).expect("setup of Tracer");
-    let otel_layer = tracing_opentelemetry::layer().with_tracer(otel_tracer);
 
     opentelemetry::global::set_text_map_propagator(
         propagation::TextMapSplitPropagator::default(),
     );
 
-    let fmt_layer = tracing_subscriber::fmt::layer()
-        .json()
-        .with_timer(tracing_subscriber::fmt::time::uptime())
-        .with_span_events(FmtSpan::NEW | FmtSpan::CLOSE)
-        .with_writer(std::io::stdout.with_max_level(log_level));
-
-    let level_filter: LevelFilter = max(log_level, tracing::Level::INFO).into();
+    let level_filter: LevelFilter = max(log_level, tracing::Level::TRACE).into();
     let subscriber = tracing_subscriber::registry()
-        .with(fmt_layer)
         .with(level_filter)
-        .with(otel_layer);
+        .with(build_logger_text(log_level));
+    let _guard = tracing::subscriber::set_default(subscriber);
```

Comment:
Why do we create `subscriber` twice? Is it intentional?

Reply:
Something like this:

```rust
    let level_filter: LevelFilter = max(log_level, tracing::Level::TRACE).into();
    let otel_layer = tracing_opentelemetry::layer().with_tracer(otel_tracer);
    let subscriber = tracing_subscriber::registry()
        .with(build_logger_text(log_level))
        .with(otel_layer)
        .with(level_filter);
    tracing::subscriber::set_global_default(subscriber).unwrap();
```

Reply:
that was the code before (without the trace) level